src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,225 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2019 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef MPIR_HW_TOPO_H_INCLUDED
+#define MPIR_HW_TOPO_H_INCLUDED
+
+/*
+ * Node hardware object types
+ */
+typedef enum {
+    MPIR_NODE_OBJ_TYPE__MACHINE,
+    MPIR_NODE_OBJ_TYPE__PACKAGE,
+    MPIR_NODE_OBJ_TYPE__CORE,
+    MPIR_NODE_OBJ_TYPE__PU,
+    MPIR_NODE_OBJ_TYPE__L1CACHE,
+    MPIR_NODE_OBJ_TYPE__L2CACHE,
+    MPIR_NODE_OBJ_TYPE__L3CACHE,
+    MPIR_NODE_OBJ_TYPE__L4CACHE,
+    MPIR_NODE_OBJ_TYPE__L5CACHE,
+    MPIR_NODE_OBJ_TYPE__L1ICACHE,
+    MPIR_NODE_OBJ_TYPE__L2ICACHE,
+    MPIR_NODE_OBJ_TYPE__L3ICACHE,
+    MPIR_NODE_OBJ_TYPE__GROUP,
+    MPIR_NODE_OBJ_TYPE__NUMANODE,
+    MPIR_NODE_OBJ_TYPE__BRIDGE,
+    MPIR_NODE_OBJ_TYPE__PCI_DEVICE,
+    MPIR_NODE_OBJ_TYPE__OS_DEVICE,
+    MPIR_NODE_OBJ_TYPE__MISC,
+    MPIR_NODE_OBJ_TYPE__MAX
+} MPIR_Node_obj_type;
+
+/*
+ * Node hardware OS device object types
+ */
+typedef enum {
+    MPIR_NODE_OBJ_OSDEV_TYPE__BLOCK,
+    MPIR_NODE_OBJ_OSDEV_TYPE__GPU,
+    MPIR_NODE_OBJ_OSDEV_TYPE__NETWORK,
+    MPIR_NODE_OBJ_OSDEV_TYPE__OPENFABRICS,
+    MPIR_NODE_OBJ_OSDEV_TYPE__DMA,
+    MPIR_NODE_OBJ_OSDEV_TYPE__COPROC
+} MPIR_Node_obj_osdev_type;
+
```

Comment:
@pavanbalaji I left this unchanged as I wasn't sure renaming it as `MPIR_Node_osdev_type` would be the best thing to do. This is currently following the `hwloc` nomenclature. In fact all objects, regardless of their type, are represented as `MPIR_Node_obj` (`hwloc_obj_t`). Maybe an alternative naming could be `MPIR_Node_osdev_obj_type`?

Reply:
I take back my comment about this change.  On reading it again, the current naming does make sense.  `MPIR_Node_obj_type` defines object types.  When we pick `MPIR_NODE_OBJ_TYPE__OS_DEVICE`, the `MPIR_Node_obj_osdev_type` enum gives more information on which OS device.  So you can leave it as it is.